hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/RingbufferService.java

```diff
@@ -347,126 +337,56 @@ public void ensureQuorumPresent(String distributedObjectName, QuorumType require
 
     @Override
     public Runnable prepareMergeRunnable() {
-        Map<Integer, List<RingbufferContainer>> partitionContainerMap = createHashMap(containers.size());
-        for (Entry<Integer, Map<ObjectNamespace, RingbufferContainer>> entry : containers.entrySet()) {
-            int partitionId = entry.getKey();
-            if (!partitionService.isPartitionOwner(partitionId)) {
-                continue;
-            }
-
-            List<RingbufferContainer> containerList = partitionContainerMap.get(partitionId);
-            if (containerList == null) {
-                containerList = new LinkedList<RingbufferContainer>();
-                partitionContainerMap.put(partitionId, containerList);
-            }
-            for (RingbufferContainer container : entry.getValue().values()) {
-                String serviceName = container.getNamespace().getServiceName();
-                // we just merge ringbuffer containers which are not used by other services
-                if (SERVICE_NAME.equals(serviceName) && !(getMergePolicy(container) instanceof DiscardMergePolicy)) {
-                    containerList.add(container);
-                }
-                container.cleanup();
-            }
-        }
-
-        // clear all items either owned or backup
-        reset();
-
-        return new Merger(partitionContainerMap);
+        RingbufferContainerCollector collector = new RingbufferContainerCollector(nodeEngine, containers);
+        collector.run();
+        return new Merger(collector);
```

Comment:
Same as in `MultiMapService`, the containers are used after they have been `reset()` but no harm here since only the collection `containers` is cleared.